master/buildbot/process/buildrequestdistributor.py

```diff
@@ -0,0 +1,1004 @@
+    # an old instance and creating a new one.
+    #
+    # The entry point is:
+    #    * bc.chooseNextBuild() - get the next (slave, [breqs]) or (None, None)
+    #
+    # The default implementation of this class implements a default
+    # chooseNextBuild() that delegates out to two other functions:
+    #   * bc.popNextBuild() - get the next (slave, breq) pair
+    #   * bc.mergeRequests(breq) - perform a merge for this breq and return
+    #       the list of breqs consumed by the merge (including breq itself)
+
+    def __init__(self, bldr, master):
+        self.bldr = bldr
+        self.master = master
+        self.breqCache = {}
+        self.unclaimedBrdicts = None
+        self.resumeBrdicts = None
+
+
+    @defer.inlineCallbacks
+    def chooseNextBuild(self):
+        # Return the next build, as a (slave, [breqs]) pair
+
+        slave, breq = yield self.popNextBuild()
+        if not slave or not breq:
+            defer.returnValue((None, None))
+            return
+        
+        breqs = yield self.mergeRequests(breq)
+        for b in breqs:
+                self._removeBuildRequest(b)
+                
+        defer.returnValue((slave, breqs))
+    
+
+    # Must be implemented by subclass
+    def popNextBuild(self):
+        # Pick the next (slave, breq) pair; note this is pre-merge, so
+        # it's just one breq
+        raise NotImplementedError("Subclasses must implement this!")
+    
+    # Must be implemented by subclass
+    def mergeRequests(self, breq):
+        # Merge the chosen breq with any other breqs that are compatible
+        # Returns a list of the breqs chosen (and should include the
+        # original breq as well!)
+        raise NotImplementedError("Subclasses must implement this!")
+    
+
+    # - Helper functions that are generally useful to all subclasses -
+    
+    @defer.inlineCallbacks
+    def _fetchUnclaimedBrdicts(self):
+        # Sets up a cache of all the unclaimed brdicts. The cache is
+        # saved at self.unclaimedBrdicts cache. If the cache already 
+        # exists, this function does nothing. If a refetch is desired, set 
+        # the self.unclaimedBrdicts to None before calling."""
+        
+        if self.unclaimedBrdicts is None:
+            brdicts = yield self.master.db.buildrequests.getBuildRequests(
+                        buildername=self.bldr.name, claimed=False)
+            # sort by submitted_at, so the first is the oldest
+            brdicts.sort(key=lambda brd : brd['submitted_at'])
+            self.unclaimedBrdicts = brdicts
+        defer.returnValue(self.unclaimedBrdicts)
+
+    @defer.inlineCallbacks
+    def claimBuildRequests(self, breqs):
+        brids = [br.id for br in breqs]
+        yield self.master.db.buildrequests.claimBuildRequests(brids)
+    
+    @defer.inlineCallbacks
+    def _getBuildRequestForBrdict(self, brdict):
+        # Turn a brdict into a BuildRequest into a brdict. This is useful 
```

Comment:
@opdude the comment seems wrong it creates a BuildRequest from a brdict and this object is cached